### ISSUE TITLE
DB index is not used on some queries

### ISSUE THREAD

**Issue by enono78** [adamchainz/django-mysql#480]
### Summary: 

django-mysql allows to perform a query on JSON encapsulated fields. To prevent performance issues, we created an index as explained [here](https://mysqlserverteam.com/indexing-json-documents-via-virtual-columns/).
So we expect the index to be used when running this query:
```
TableDoc.objects.filter(field__encaspulated_field="value")
```
But it does not, as you can see below (the sql query is the one generated by django):
```
EXPLAIN SELECT * FROM `tabledoc` 
WHERE JSON_EXTRACT(`tabledoc`.`field`, '$.encaspulated_field') = (CAST('\"value\"' AS JSON));
```
id|	select_type|	table|	partitions|	type|possible_keys|	key|	key_len|	ref|	rows|	filtered|	Extra
----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |  ----- | ----- | ----- | ----- |  
1|SIMPLE	|tabledoc	|NULL	|ALL|	encaspulated_field|	NULL|	NULL|	NULL|	71763|	10.00|	Using where; Using filesort

**When modifying the query to use naively the value without JSON CAST, the index is well used.**
```
EXPLAIN SELECT * FROM `tabledoc` 
WHERE JSON_EXTRACT(`tabledoc`.`field`, '$.encaspulated_field') = "value");
```
id|	select_type|	table|	partitions|	type|possible_keys|	key|	key_len|	ref|	rows|	filtered|	Extra
----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |  ----- | ----- | ----- | ----- |  
1|SIMPLE	|tabledoc	|NULL	|ALL|	encaspulated_field|	NULL|	NULL|	NULL|	1|	10.00|	Using where; Using filesort

I'm not sure if it is a bug and if it has been solved on more recent django/django-mysql versions but I would appreciate your feedbacks.

Thanks :)

### Configuration

Django Version: *Django 1.9.1*

Database and version used: *MySQL 5.7*

Version: *Django-MySQL 1.1.0*

**Comment by adamchainz** [comment]
I'd consider this a performance bug. I don't believe anything has changed on this in later versions of django-mysql, but there isn't a reason not to upgrade - 2.0.0 was only triggered by two small backwards-incompatible changes that probably don't affect you: https://github.com/adamchainz/django-mysql/blob/master/HISTORY.rst#200-2017-05-28 . Also I hope you find time to upgrade to the latest Django 1.9.x at least, there are lots of security fixes you're missing out on!

As for solving this, perhaps we can improve the search behaviour to not always insert `CAST` on the right hand side for expressions that resolve in simple values. Would you be interested in looking into this?

**Comment by enono78** [comment]
Hi @adamchainz,
I gave a short shot by forking the repository and change some stuff in the `JSONValue` class but it was not enough... I will try harder to find a solution. Thanks you.

**Comment by adamchainz** [comment]
@enono78 do you want to come back to this?

**Comment by adamchainz** [comment]
This has been fixed in #621 and released in version 3.4.0.
